### PR TITLE
Ctl cleanup

### PR DESCRIPTION
**Pull request by PatKamin**
Requires: #608 
### Types of changes

- [ ] Bugfix (non-breaking change which fixes issue linked in Description above)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation (correction or otherwise)
- [ ] Cosmetics (whitespace, appearance)
- [ ] Other

### Checklist

- [ ] Code compiles without errors
- [ ] All tests pass locally with my changes (see TESTING section in CONTRIBUTING file)
- [ ] Created tests which will fail without the change (if possible)
- [ ] Extended the README/documentation (if necessary)
- [ ] All newly added files have proprietary license (if necessary)
- [ ] All newly added files are referenced in MANIFEST files (if necessary)

## Further comments

---
This change is [Reviewable](https://reviewable.io/reviews/memkind/memkind/612)

### CODE REVIEW

File: tiering/ctl.c [hunk 1]
```diff
@@ -376,6 +366,14 @@ struct memtier_kind *ctl_create_tier_kind_from_env(char *env_var_string)
         }
     }
 
+    ret = ctl_parse_policy(qbuf, &policy);
+    if (ret != 0) {
+        log_err("Failed to parse policy: %s", qbuf);
+        memkind_free(MEMKIND_DEFAULT, tier_cfgs);
```
Comment (michalbiesek) [inline tiering/ctl.c:373]: goto builder_delete;

Comment (PatKamin) [reply tiering/ctl.c]: Done.

File: tiering/ctl.c [hunk 2]
```diff
@@ -199,37 +199,62 @@ static int ctl_parse_ratio(char **sptr, unsigned *dest)
...
         }
 
-        ret = ctl_parse_pmem_size(&sptr, &tier_cfg->pmem_size);
+        ret = ctl_parse_pmem_size(&sptr, &pmem_size);
         if (ret != 0) {
             return -1;
         }
     }
 
+    *kind = ctl_get_kind(kind_name, pmem_path, pmem_size);
```
Comment (michalbiesek) [inline tiering/ctl.c:256]: Can this now be removed (last commit)?
https://github.com/memkind/memkind/blob/c58b8145fb47b3499cfe86966d954fd23fa70611/tiering/ctl.c#L23-L24

Comment (PatKamin) [reply tiering/ctl.c:246]: Done.

**Review by bratpiorka — COMMENTED** [review]
> Reviewed 1 of 1 files at r1, 1 of 1 files at r2, 1 of 1 files at r3, 1 of 1 files at r4, 1 of 1 files at r5, 1 of 1 files at r6, 1 of 1 files at r10, 1 of 1 files at r11, 1 of 1 files at r12.
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/612)* status: all files reviewed, 3 unresolved discussions (waiting on @michalbiesek and @PatKamin)

___
*[tiering/ctl.c, line 18 at r11](https://reviewable.io/reviews/memkind/memkind/612#-MXCS5Cg2X5raC00qRKc:-MXCS5Cg2X5raC00qRKd:bfn9hpk) ([raw file](https://github.com/memkind/memkind/blob/819f1c41a839d76acfb0c2922433f36fe46e918f/tiering/ctl.c#L18)):*
> ```C
> #define CTL_STRING_QUERY_SEPARATOR ","
> 
> #define MAX_TIERS 64
> ```

Is this limitation described somewhere?

**Review by PatKamin — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/612)* status: all files reviewed, 3 unresolved discussions (waiting on @bratpiorka and @michalbiesek)

___
*[tiering/ctl.c, line 18 at r11](https://reviewable.io/reviews/memkind/memkind/612#-MXCS5Cg2X5raC00qRKc:-MXCSkCd6FqFQRCHel_p:b-tfkman) ([raw file](https://github.com/memkind/memkind/blob/819f1c41a839d76acfb0c2922433f36fe46e918f/tiering/ctl.c#L18)):*
<details><summary><i>Previously, bratpiorka (Rafał Rudnicki) wrote…</i></summary><blockquote>

Is this limitation described somewhere?
</blockquote></details>

No, it's not. This limitation is going to be lifted during this cleanup. We can't have two limits for the same thing (in API unsigned is the limit).

**Review by bratpiorka — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/612)* status: all files reviewed, 2 unresolved discussions (waiting on @michalbiesek)

___
*[tiering/ctl.c, line 18 at r11](https://reviewable.io/reviews/memkind/memkind/612#-MXCS5Cg2X5raC00qRKc:-MXCTRS--5ryYB9H6nYS:b32wtd) ([raw file](https://github.com/memkind/memkind/blob/819f1c41a839d76acfb0c2922433f36fe46e918f/tiering/ctl.c#L18)):*
<details><summary><i>Previously, PatKamin (Patryk Kaminski) wrote…</i></summary><blockquote>

No, it's not. This limitation is going to be lifted during this cleanup. We can't have two limits for the same thing (in API unsigned is the limit).
</blockquote></details>

ok

**Comment by michalbiesek** [comment]
LGTM so far, maybe as next step you could try to put strcmp against name from env in one place,:
see that currently it is placed in ctl_validate_kind_name, ctl_get_kind, but IMHO it could be placed only here:
https://github.com/PatKamin/memkind/blob/7c801c7991c250671b25c2aff1baf5a8966a84ae/tiering/ctl.c#L234-L244

**Review by michalbiesek — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/612)* status: 0 of 1 files reviewed, 2 unresolved discussions (waiting on @bratpiorka and @PatKamin)

___
*[tiering/ctl.c, line 221 at r13](https://reviewable.io/reviews/memkind/memkind/612#-MXCwphX4gUgyQjug-mu:-MXCwphX4gUgyQjug-mv:bo8xmei) ([raw file](https://github.com/memkind/memkind/blob/5d1f41ea6a9c27870fc6e98fd1a94d3c08960b03/tiering/ctl.c#L221)):*
> ```
> if (strcmp(kind_name, "DRAM") && strcmp(kind_name, "FS_DAX"))
> ```
<details><summary><i alt="summary">Quoted 9 lines of code…</i></summary>

> ```
> if (strcmp(kind_name, "DRAM") {
>       *kind = MEMKIND_DEFAULT;
>       log_debug("kind_name: memkind_default");
>)else if (!strcmp(kind_name, "FS_DAX")) {
>   (actions from memkind_t ctl_get_kind...)
>} else {
>        log_err("Unsupported kind: %s", kind_name);
>        return -1;
> }
> ```
</details>

___
*[tiering/ctl.c, line 236 at r13](https://reviewable.io/reviews/memkind/memkind/612#-MXCwk5I0n6raevuDD3S:-MXCwk5I0n6raevuDD3T:br7sdge) ([raw file](https://github.com/memkind/memkind/blob/5d1f41ea6a9c27870fc6e98fd1a94d3c08960b03/tiering/ctl.c#L236)):*
> ```
> ctl_get_kind(kind_name, pmem_path, pmem_size);
> ```
you could get rid of this method - see my other comment

**Review by PatKamin — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/612)* status: 0 of 2 files reviewed, 2 unresolved discussions (waiting on @bratpiorka and @michalbiesek)

___
*[tiering/ctl.c, line 221 at r13](https://reviewable.io/reviews/memkind/memkind/612#-MXCwphX4gUgyQjug-mu:-MXCz-2C-Te3VEiHTykT:b-896fix) ([raw file](https://github.com/memkind/memkind/blob/5d1f41ea6a9c27870fc6e98fd1a94d3c08960b03/tiering/ctl.c#L221)):*
<details><summary><i>Previously, michalbiesek wrote…</i></summary><blockquote>

> ```
> if (strcmp(kind_name, "DRAM") && strcmp(kind_name, "FS_DAX"))
> ```
> ```
> if (strcmp(kind_name, "DRAM") {
>       *kind = MEMKIND_DEFAULT;
>       log_debug("kind_name: memkind_default");
>)else if (!strcmp(kind_name, "FS_DAX")) {
>   (actions from memkind_t ctl_get_kind...)
>} else {
>        log_err("Unsupported kind: %s", kind_name);
>        return -1;
> }
> ```
</blockquote></details>

Done.
___
*[tiering/ctl.c, line 236 at r13](https://reviewable.io/reviews/memkind/memkind/612#-MXCwk5I0n6raevuDD3S:-MXCz5qMDnAUptH0wVr1:b-896fix) ([raw file](https://github.com/memkind/memkind/blob/5d1f41ea6a9c27870fc6e98fd1a94d3c08960b03/tiering/ctl.c#L236)):*
<details><summary><i>Previously, michalbiesek wrote…</i></summary><blockquote>

> ```
> ctl_get_kind(kind_name, pmem_path, pmem_size);
> ```
you could get rid of this method - see my other comment
</blockquote></details>

Done.

**Review by michalbiesek — APPROVED** [review]
> Reviewed 1 of 2 files at r14.
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/612)* status: 0 of 2 files reviewed, all discussions resolved (waiting on @bratpiorka and @michalbiesek)

**Review by michalbiesek — COMMENTED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/612#-:-MXaVCH--6tmFEJtucaN:bnfp4nl)* status: 0 of 2 files reviewed, all discussions resolved (waiting on @bratpiorka and @michalbiesek)

**Review by bratpiorka — COMMENTED** [review]
:lgtm:

> Reviewed 1 of 2 files at r14, 1 of 1 files at r15.
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/612#-:-MXaWTtkEE0yScgbxa2T:bnfp4nl)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @PatKamin)

**Pull request merged**